app/models/user.rb

```diff
@@ -1,19 +1,33 @@
 class User < ActiveRecord::Base
-  #Validations
-  validates :username,:email,  presence: true
+  # Associations
+  has_many :tweets
+  has_many :grams, through: :instagram_users
+  has_many :tweets, through: :tw_users
+  has_and_belongs_to_many :instagram_users
+  has_and_belongs_to_many :tw_users
 
- #Methods
+  # Validations
+  validates :avatar_url,:uid,:provider,  presence: true
+  validates :username, presence: true, uniqueness: true
 
-  #  def self.find_or_create_from_omniauth(auth_hash)
+  # Methods
+  def self.find_or_create_from_omniauth(auth_hash)
+    instagram = auth_hash["user"]
+    auth_uid = instagram.nil? ? "nil" : instagram["id"]
+    auth_provider = instagram.nil? ? "developer" : "instagram"
 
-  #   auth_uid = auth_hash["uid"]
-  #   auth_provider = auth_hash["provider"]
+    user = User.where(uid: auth_uid, provider: auth_provider).first_or_initialize
+    user.uid      = instagram.nil?   ? "nil" : instagram["id"]
```

Comment:
Why use a text value "nil" instead of an actual nil? Databases generally support nullable types.